pkg/engine/plan/type_field_extractor.go

```diff
@@ -0,0 +1,197 @@
+package plan
+
+import (
+	"github.com/jensneuse/graphql-go-tools/pkg/ast"
+	"github.com/jensneuse/graphql-go-tools/pkg/astparser"
+	"github.com/jensneuse/graphql-go-tools/pkg/astprinter"
+	"github.com/jensneuse/graphql-go-tools/pkg/asttransform"
+	"github.com/jensneuse/graphql-go-tools/pkg/federation"
+)
+
+const (
+	federationKeyDirectiveName      = "key"
+	federationRequireDirectiveName  = "requires"
+	federationExternalDirectiveName = "external"
+)
+
+// TypeFieldExtractor takes an ast.Document as input
+// and generates the TypeField configuration for both root fields & child fields
+// If a type is a federation entity (annotated with @key directive)
+// and a field is is extended, this field will be skipped
+// so that only "local" fields will be generated
+type TypeFieldExtractor struct {
+	document *ast.Document
+}
+
+func NewNodeExtractor(document *ast.Document) *TypeFieldExtractor {
+	return &TypeFieldExtractor{document: document}
+}
+
+// GetAllNodes returns all Root- & ChildNodes
+func (r *TypeFieldExtractor) GetAllNodes() (rootNodes, childNodes []TypeField) {
+	rootNodes = r.getAllRootNodes()
+	childNodes = r.getAllChildNodes(rootNodes)
+	return
+}
+
+func (r *TypeFieldExtractor) getAllRootNodes() []TypeField {
+	var rootNodes []TypeField
+
+	for _, astNode := range r.document.RootNodes {
+		switch astNode.Kind {
+		case ast.NodeKindObjectTypeExtension, ast.NodeKindObjectTypeDefinition:
+			r.addRootNodes(astNode, &rootNodes)
+		}
+	}
+
+	return rootNodes
+}
+
+func (r *TypeFieldExtractor) getAllChildNodes(rootNodes []TypeField) []TypeField {
+	var childNodes []TypeField
+
+	for i := range rootNodes {
+		fieldNameToRef := make(map[string]int, len(rootNodes[i].FieldNames))
+
+		rootNodeASTNode, exists := r.document.Index.FirstNodeByNameStr(rootNodes[i].TypeName)
+		if !exists {
```

Comment:
should we check node Kind here?

Reply:
it's not required, NodeFieldDefinition does this already